docs/defaulting.md

```diff
@@ -0,0 +1,180 @@
+service is a standalone component in KubeVirt and can then be integrated into
+two different flows:
+
+ 1. When doing a POST to
+    `/apis/kubevirt.io/v1alpha1/namespaces/mynamespace/vms`, `virt-api` can do
+    a roundtrip to that service and prefill the defaults. As the response of
+    the POST to `virt-api`, you will get the completely prefilled VM
+    specification. This updated spec contains all the necessary bits, to allow
+    consistent hotplugging and keeping the VM definition consistent between
+    redefines.
+ 2. Add an extra endpoint to `virt-api` which only purpose is to prefill the VM
+    spec. The URI will be `/apis/kubevirt.io/v1.alpha1/vms/defaults`
+    (non-namespaced). You can POST VM specifications there and get a fully
+    populated VM spec back. With this service URL in place, mixing
+    administrative tasks like hotplug, adding devices for later runs and
+    changing the spec for later runs (e.g. memory) can be solved in a
+    consistent way, without the need to ever do an asynchronous roundtrip to
+    libvirt to build consistent specs. **This allows decoupling administrative
+    tasks completely from the runtime**. 
+
+Defaulting rules:
+
+ * The Kubevirt VM spec can contain abstract definitions, like a reference to a
+   volume claim. In such cases most of the time, the device `source` section
+   needs to be replaced by a dummy `source` inside the defaulter. Then the
+   fully populated `target` section needs to be mapped back.
+ * There exist mandatory Domain XML fields, which we can't allow to be set on
+   the VM specification. In such cases, the defaulter needs to set dummy values
+   before asking Libvirt for the defaults. Since these fields should not exist
+   on the VM specification at all (see
+   [docs/vm-configuration.md](docs/vm-configuration.md)), not mapping such
+   dummy data back should be easy.
+
+## Important side notes
+
+### Device naming
+
+With the proposed Defaulting service, it is still pretty hard, from an
+administration and operations perspective to find out which device is which.
+Libvirt at the moment does not support naming devices for the whole VM
+lifecycle (https://bugzilla.redhat.com/show_bug.cgi?id=1434451). It actually
+becomes a bigger problem with the defaulting service in place. To find out if a
+device has changed, or needs an update, you would have to rely on the order of
+the devices (which is not necessarily stable:
+https://www.redhat.com/archives/libvirt-users/2012-December/msg00087.html), or
+try to match them based on their content.
+
+### Migrating between different QEMU versions
+
+In order to make sure that the Libvirt version, which does the defaulting
+generates appliable defaults, it must first be compatible with the libvirt/qemu
+pairs, currently running on the host. Second a guest machine type must be
+specified, to guarantee migratability between the different versions in the
+cluster.
```

Comment:
First I think it is outside of this proposal, since we don't guarantee that right now. Second, I think, having something like a configmap, which contains for instance a namespace or cluser wide guest machine would do the trick. Like you do it right now in oVirt. You tell your cluster up front what to use for these values. If we want to enforce this or not, depends if we want to allow creating cluster which do or don't care about migrations.